fix: missing `MsgModifyBid` handler (#143)
tendermint/fundraising: docs/How-To/cli/README.md

---
# CLI Reference

## Synopsis

This document provides a high-level overview of how the command line (CLI) interface works for the `fundraising` module. To set up a local testing environment, it requires the latest [Starport](https://starport.com/). If you don't have Starport set up in your local machine, see [this Starport guide](https://docs.starport.network/) to install it. Run this command under the project root directory `$ starport chain serve -c config-test.yml` or simply `$ make localnet`.

Note that [jq](https://stedolan.github.io/jq/) is recommended to be installed as it is used to process JSON throughout the document.

## Command Line Interface

- [Transaction](#transaction)
  - [CreateFixedPriceAuction](#createfixedpriceauction)
  - [CreateBatchAuction](#createbatchauction)
  - [CancelAuction](#cancelauction)

>   - [AddAllowedBidder](#addallowedbidder)

Reviewer: We defined AddAlloedBidder that can be called from the outside of the module. Is there any difference from the other modules?

Author: `AddAllowedBidder` CLI command is only available for the testing purpose. It becomes available when `fundraisingd` is built with `config-test.yml` file which passes custom ldflags to keeper. Without `AddAllowedBidders` CLI command, there is currently no way for developers to test out `PlaceBid` or `ModifyBid`.

create-batch-auction [file]
```

| **Argument** |  **Description**                                             |
| :----------- | :----------------------------------------------------------- |

> | file         | file that contains required fields for a fixed price auction |

Reviewer: ```suggestion
| file         | file that contains required fields for a batch auction |
```

Author: Thanks for the catch!

Field description of the input file

| **Field**           |  **Description**                                                                    |
| :------------------ | :---------------------------------------------------------------------------------- |

> | start_price         | The starting price of the selling coin; it is proportional to the paying coin denom. This is the matched price | 

Reviewer: Since this is not the matched price, how about modifying like the following?
```suggestion
| start_price         | The starting price of the selling coin; it is proportional to the paying coin denom. This might not be the matched price and this can be used as a reference of the price from the auctioneer. | 
```

Author: Yes, `This is the matched price` is incorrect description. I will update it with the suggested change.

| min_bid_price       | The minimum bid price that bidders must place with                                  |
| selling_coin        | The selling amount of coin for the auction                                          | 
| paying_coin_denom   | The paying coin denom that bidders use to bid with                                  | 
| vesting_schedules   | The vesting schedules that release the paying coins to the autioneer                | 
| max_extended_round  | The maximum number of extended rounds that provides additional opportunity for the bidders to place bids when more than a certain ratio of the number of the matched bids are reduced compared to the previous end time  |
| extended_round_rate | The threshold reduction of the number of the matched bids are reduced compared to the previous end time to decide the necessity of another extended round | 
| start_time          | The start time of the auction                                                       | 
| end_time            | The end time of the auction                                                         | 

Example of input as JSON:

> An auctioneer creates a batch price auction for `1000000000000denom1` selling coin where the start price is `0.3` and the minimum bid price is `0.1`. Unlike a fixed price auction, the start price is not determined until the auction ends. It is only used as a reference of what the auctioneer considers their selling coin price is. The auctioneer sets their vesting schedules for themselves to receive the accumulated paying coin amount. This is a gesture for taking responsibility for auction participants.

Reviewer: ```suggestion
An auctioneer creates a batch price auction for `1000000000000denom1` selling coin where the start price is `0.3` and the minimum bid price is `0.1`. Unlike a fixed price auction, the start price might not be the matched price. It is only used as a reference of what the auctioneer considers their selling coin price is. The auctioneer sets their vesting schedules for themselves to receive the accumulated paying coin amount. This is a gesture for taking responsibility for auction participants.
```

--broadcast-mode block \
--yes \
--output json | jq

# Place a batch-worth bid type for the batch auction
fundraisingd tx fundraising bid 1 batch-worth 0.35 10000000denom2 \
--chain-id fundraising \
--from bob \
--keyring-backend test \
--broadcast-mode block \
--yes \
--output json | jq

# Place a batch-many bid type for the batch auction
fundraisingd tx fundraising bid 1 batch-many 0.4 10000000denom1 \
--chain-id fundraising \
--from bob \
--keyring-backend test \
--broadcast-mode block \
--yes \
--output json | jq

#
# Tips
#
# Query all bids that belong to the auction
fundraisingd q fundraising bids 1 -o json | jq
```

## ModifyBid

This command is used for modifying the bid. It is only supported for `BatchAuction`. The bidder is allowed to modify the bid only with the same bid type and they must provide either higher bid price or larger bid amount. Lowering bid price or lesser bid amount is restricted.

Usage

```bash
modify-bid [auction-id] [bid-id] [price] [coin]
```

| **Argument**|  **Description**              |
| :---------- | :---------------------------- |
| auction-id  | auction id                    | 
| bid-id      | bid id that the bidder placed |

> | price       | bid price of a selling coin as the unit of a paying coin. For `BatchAuction`, The price cannot be lower than the original bid price of the auction | 

Reviewer: ```suggestion
| price       | bid price of a selling coin as the unit of a paying coin. For `BatchAuction`, the price cannot be lower than the original bid price of the auction | 
```

> | price       | bid price of a selling coin as the unit of a paying coin. For `FixedPriceAuction`, it must be the start price of the auction. For `BatchAuction`, the price must be higher than or equal to the minimum bid price of the auction | 

Reviewer: ```suggestion
| price       | bid price of a selling coin as the unit of a paying coin. The price cannot be lower than the original bid price of the auction | 
```

Author: Should we not need to explain about price for `FixedPriceAuction`?

Author: How about the following change?

> bid price of a selling coin as the unit of a paying coin. For `FixedPriceAuction`, the price must be the same as start price and for `BatchAuction`, the price cannot be lower than the original bid price of the auction.

Reviewer: Since `FixedPriceAuction` does not have `ModifyBid`, I don't think we need to explain this for `FixedPriceAuction`. If need, how about mentioning that `FixedPriceAuction` does not support to modify the bid previously placed?

Author: Oh I misunderstood this line with `PlaceBid`. I get it now. Let me go make the change.

> | coin        | how many coins to bid. The denom must be the same as the modifying bid |

Reviewer: ```suggestion
| coin        | how many coins to bid. The denom must be the same as the modifying bid. The amount cannot be smaller than that of the original coin amount. |
```
